hybridse/src/udf/default_defs/window_functions_def.cc

```diff
@@ -90,6 +94,112 @@ node::ExprNode* BuildAt(UdfResolveContext* ctx, ExprNode* input, ExprNode* idx,
+                .init(prefix + "_init", Init)
+                .update(prefix + "_update", Update)
+                .output(prefix + "_output", reinterpret_cast<void*>(Output), true);
+        }
+
+     private:
+        struct ContainerT {
+            ContainerT() : nth(0), cur_idx(0) {}
+
+            // given nth value, be negative or positive
+            NthType nth;
+
+            NthType cur_idx;
+
+            // saved value list
+            // if `nth` > 0, work like a ring buffer
+            // if `nth < 0`, have one value at most
+            std::queue<std::pair<T, bool>, std::list<std::pair<T, bool>>> data;
+        };
+
+        // (nth idx, [(value, value_is_null)])
+        // using ContainerT = std::pair<NthType, std::vector<std::pair<T, bool>>>;
+
+        static void Init(ContainerT* ctr) { new (ctr) ContainerT(); }
+
+        static ContainerT* Update(ContainerT* ctr, CType value, bool value_is_null, NthType nth, bool cond,
+                                  bool cond_is_null) {
+            if (nth == 0) {
+                return ctr;
+            }
+
+            // by default, update is iterated from window end to window start
+            if (ctr->nth == 0) {
+                // update only once, we treat nth as constant
+                ctr->nth = nth;
+            }
+
+            if (cond && !cond_is_null) {
+                if (ctr->nth > 0) {
+                    // nth from window start
+                    ctr->data.emplace(container::ContainerStorageTypeTrait<T>::to_stored_value(value), value_is_null);
+                    if (ctr->data.size() > ctr->nth) {
+                        ctr->data.pop();
+                    }
+                } else {
+                    // nth from window end
+                    ctr->cur_idx++;
+                    if (ctr->cur_idx + ctr->nth == 0) {
+                        // reaches nth
+                        ctr->data.emplace(container::ContainerStorageTypeTrait<T>::to_stored_value(value),
+                                           value_is_null);
+                    }
+                }
+            }
+            return ctr;
+        }
+
+        static void Output(ContainerT* ctr, T* value, bool* is_null) {
+            absl::Cleanup clean = [ctr]() { ctr->~ContainerT(); };
+
+            if (ctr->nth == 0) {
+                *is_null = true;
+                return;
+            }
+
+            if (ctr->nth > 0) {
+                // count from window start to window end
+                if (ctr->data.size() < ctr->nth) {
+                    // window size < nth -> NULL
+                    *is_null = true;
+                    return;
+                }
+                *value = ctr->data.front().first;
+                *is_null = ctr->data.front().second;
+                return;
+            }
+
+            // count from window end
+            if (ctr->data.empty()) {
```

Comment:
it may be more simply as below: 
```
if (ctr->nth == 0 || ctr->data.empty() || (ctr->nth > 0 && ctr->data.size() < ctr->nth)) {
...
}
*value = ctr->data.front().first;
*is_null = ctr->data.front().second;
```